### ISSUE TITLE
Enabling dynamic and fixed indices.

### ISSUE THREAD

**Issue by Tokazama** [andyferris/Dictionaries.jl#5]
This came up on [discourse](https://discourse.julialang.org/t/ann-dictionaries-jl-improved-productivity-and-performance-of-dictionaries-in-julia/32074/16?u=zach_christensen). It's pretty important to have indices that are mutable for a lot of things methods to work (append!, push!, etc.), but there are times it's worth improving performance and losing mutating functions. I mentioned my work on [StaticRanges.jl](https://github.com/Tokazama/StaticRanges.jl) may be helpful. It provides mutable ranges as well with protective `setproperty!` methods so that users don't accidently make bizarre changes to a mutable range.

This could be as simple as:

```julia
const StaticIndicesRange{T,L} = Indices{T,OneToSRange{T,L}}
const DynamicIndicesRange{T} = Indices{T,OneToMRange{T}}
const FixedIndicesRange{T} = Indices{T,OneTo{T}}
```

If this is something you're interested in I'd be interested in further discussing how to implement it.

**Comment by andyferris** [comment]
Yes, once some ordering semantics of `AbstractDictionary` are resolved in #13 I don't see why we can't have array-like dictionaries with ranges as keys. 

I don't think implementing this will be too hard - e.g. `Indices` can have tokens and lookup can defer to `findfirst` for `AbstractArray`s, we can set the `issettable` and `isinsertable` traits accordingly, or something like that... hmm...

**Comment by Tokazama** [comment]
I'd ultimately like to incorporate this package into [AxisIndices.jl](https://github.com/Tokazama/AxisIndices.jl) so once #13 is in place I'd be happy to help out with the details of this.
